### DISCUSSION TITLE
JSON

### DISCUSSION THREAD

**Discussion by fitneva** [jspsych/jsPsych#885] [Q&A] [answered]
Hi, 

JSON was mentioned a couple of times and I see it stands for "JavaScript Object Notation." 

What is it/ what role does it play in jsPsych / when is it important to consider it? Not sure what the right question is.

Thanks!

**Comment by vijaymarupudi** [comment, marked as answer]
Just like how `.csv` is a data storage format (comma separated values), JSON is another storage format (more info here: <https://www.json.org/json-en.html>). `csv` is designed for tabular data, and is a bit clunky when working with other types of data. JSON in particular is flexible enough to represent many other types of data.

The syntax of JSON follows the format of a JavaScript Object or a Python Dictionary, which is why it's called JSON. Most programming languages have the ability to load, process, and store JSON.

jsPsych uses strings of JSON whenever nested data is needed (mostly so that it can be converted to `.csv` cleanly). Learning how to manipulate JSON (or Javascript objects in general) will be important when it comes to complex experiments that depend on previous trials and many other things!

The `JSON.parse()` and the `JSON.stringify()` functions will come to handy to convert JSON strings to Objects and vice-versa.

**Comment by becky-gilbert** [comment]
Hi @fitneva!

Just adding a little more to @vijaymarupudi's response:

There's an explanation in another discussion thread [here](https://github.com/jspsych/jsPsych/discussions/693#discussioncomment-38392) about how JSON can be used to take CSV-formatted trial information and turn it into a JavaScript variable, so that you can use that information in your jsPsych script.

And JSON is often used to send jsPsych data back to the server. jsPsych provides a built-in function for converting your jsPsych data into JSON format (see docs [here](https://www.jspsych.org/core_library/jspsych-data/#json)). jsPsych also provides a method for converting the data from JavaScript into CSV (see docs [here](https://www.jspsych.org/core_library/jspsych-data/#csv)), but in this case you're converting the JavaScript data to a very different (table-based) format before saving it, which means there's a small risk of losing some information if that conversion isn't straightforward. 

And FYI, there are a few differences between JSON strings and JavaScript, e.g. in JSON you can't include comments or functions as object values, and object keys need to be in double quotes: https://dev.to/desoga/javascript-object-vs-json-demystified-494j

Hope this helps! Let us know if you have other questions.
